python/dazl/client/bots.py

```
@@ -0,0 +1,315 @@
+        """
+        try:
+            self._queue = Queue()
+            while self._state not in (BotState.STOPPING, BotState.STOPPED):
+                # the main queue contains either events we have not yet processed yet or ``None``
+                # markers that merely indicate running status should be "re-checked"
+                invocation = await self._queue.get()
+                if invocation is not None:
+                    try:
+                        fut = ensure_future(self._handle_event(invocation.event))
+                        propagate(fut, invocation.future)
+                        await invocation.future
+                    except Exception:
+                        LOG.exception('An event handler in a bot has thrown an exception!')
+            self._queue = None
+        except:
+            LOG.exception('A bot thread died abnormally.')
+
+    async def _handle_event(self, event: 'BaseEvent'):
+        for event_key in EventKey.from_event(event):
+            LOG.debug('Party %s dispatching event %r to its bots...', self.party, event_key)
+            handlers = self._handlers.get(event_key)
+            if handlers is not None:
+                for handler in handlers:
+                    try:
+                        if handler.filter is None or handler.filter(event):
+                            await handler.callback(event)
+                    except Exception:
+                        LOG.exception('An event handler in a bot has thrown an exception!')
+
+    def notify(self, event: 'BaseEvent') -> 'Awaitable[None]':
+        """
+        Notifies handler(s) associated with this bot that the given event has occurred. Note that
+        this notification is asynchronous: in other words, event handlers will not have processed
+        this event by the time this function returns.
+
+        :param event: The event to raise.
+        """
+        if not isinstance(event, BaseEvent):
```

There has to be a standard library function for this idiom...

---

The standard Python idiom would be to let it fail (not even _do_ a check), but because this is async code and the actual failure would happen far from the call site, I do this check proactively.